File: bollinger/strategies.py
from . import bands

# each of these classes represent independent investment strategies

# invest if on the last 20 days close price passes at least once over upper band
class UpOnce:
    name = "uponce"
    
    def __init__(self, bands):
        self.bands = bands
    
    def invest(self):
        result = self.bands.all()
        
        for b in result:
            if b.get("close") >= b.get("up"):
                return True
        
        return False

# invest if on the last 20 days close price passes at least once under lower band
class DownOnce:
    name = "downonce"
    
    def __init__(self, bands):
        self.bands = bands
    
    def invest(self):
        result = self.bands.all()
        
        for b in result:
            if b.get("close") <= b.get("down"):
                return True
        
        return False

# invest if on the last 20 days the number of times close price passes
# over the uppper band is greater than the number of times close price passes
# under the lower band
class MoreUp:
    name = "moreup"
    
    def __init__(self, bands):
        self.bands = bands
    
    def invest(self):
        result = self.bands.all()
        
        up = 0
        down = 0
        for b in result:
            if b.get("close") >= b.get("up"):
                up = up+1
                
            if b.get("close") <= b.get("down"):
                down = down+1
        
        #print("up:%d, down: %d"%(up, down))
        
        if up > down:
            return True
        else:
            return False

# invest if on the last 20 days the number of times close price passes
# under the lower band is greater than the number of times close price passes
# over the uppper band
class MoreDown:
    name = "moredown"
    
    def __init__(self, bands):
        self.bands = bands
    
    def invest(self):
        result = self.bands.all()
        
        up = 0
        down = 0
        for b in result:
            if b.get("close") >= b.get("up"):
                up = up+1
                
            if b.get("close") <= b.get("down"):
                down = down+1
        
        #print("up:%d, down: %d"%(up, down))
        
        if down > up:
            return True
        else:
            return False

File: go/src/bollinger/strategies/strategies.go
package strategies

// several investment strategies

import (
    "bollinger/bands"
)

// invest if on the last 20 days the number of times close price passes
// under the lower band is greater than the number of times close price passes
// over the uppper band
func MoreDown(all []bands.Band) (bool) {
    up := 0
    down := 0
    
    for _,b := range(all){
        if b.Close >= b.Up {
            up++
        } else if b.Close <= b.Down{
            down++
        }
    }
    
    if down > up {
        return true
    } else {
        return false
    }
}

// invest if on the last 20 days the number of times close price passes
// over the uppper band is greater than the number of times close price passes
// under the lower band
func MoreUp(all []bands.Band) (bool) {
    up := 0
    down := 0
    
    for _,b := range(all){
        if b.Close >= b.Up {
            up++
        } else if b.Close <= b.Down{
            down++
        }
    }
    
    if up > down {
        return true
    } else {
        return false
    }
}

// invest if on the last 20 days close price passes at least once over upper band
func UpOnce(all []bands.Band) (bool) {
    for _,b := range(all){
        if b.Close >= b.Up {
            return true
        }
    }
    return false
}

// invest if on the last 20 days close price passes at least once under lower band
func DownOnce(all []bands.Band) (bool) {
    for _,b := range(all){
        if b.Close <= b.Down {
            return true
        }
    }
    return false
}
